### DISCUSSION TITLE
Execute python code when a browser session ends.

### DISCUSSION THREAD

**Discussion by mtanco** [h2oai/wave#2191] [Q&A] [answered]
I have a shared application where users can import a file for use with generative ai tasks. When they are done with the app, I would like the delete the file. Ideally, I would have a function (with access to `q`) which could run on `2023/11/10 15:42:14 # {"addr":"[::1]:49266","t":"ui_drop"}` events. Today, we get this notification in the Wave Server logs, but not in our python code at all. 

I could maybe use the `on_shutdown` function and do one big clean up when the app is killed, but this doesn't have access to `q` and also we're then paying for storage of unneeded files until we pause / kill the app. 

Help!

**Comment by mtanco** [comment, marked as answer]
```python
@on("@system.client_disconnect")
async def on_client_disconnect(q: Q):
    print("DISCONNECTED")
```

I struggled to find this in docs, but found it in discussions!

**Reply by mturoci** [reply]
Glad you found it and proved the point of having centralized public knowledge base at hand :)

The feature is documented [here](https://wave.h2o.ai/docs/routing/#handling-client-browser-tab-disconnect), but feel free to suggest a more discoverable location.
